core/src/HIP/Kokkos_HIP.cpp

```diff
@@ -50,6 +50,43 @@ void HIP::impl_initialize(InitializationSettings const& settings) {
   KOKKOS_IMPL_HIP_SAFE_CALL(
       hipGetDeviceProperties(&Impl::HIPInternal::m_deviceProp, hip_device_id));
   KOKKOS_IMPL_HIP_SAFE_CALL(hipSetDevice(hip_device_id));
+
+  // Check that we are running on the expected architecture
+  auto check_architecture = [](const std::string& gfx_name) {
+    std::string arch_name = Impl::HIPInternal::m_deviceProp.gcnArchName;
+    if (arch_name != gfx_name) {
+      std::string error_message =
+          "Kokkos::HIP::initialize ERROR: running kernels compiled for " +
+          gfx_name + " on " + arch_name + " device.\n";
+      Kokkos::abort(error_message.c_str());
+    }
+  };
+
+#ifdef KOKKOS_ARCH_AMD_GFX906
+  check_architecture("gfx906");
+#endif
+#ifdef KOKKOS_ARCH_AMD_GFX908
+  check_architecture("gfx908");
+#endif
+#ifdef KOKKOS_ARCH_AMD_GFX90A
+  check_architecture("gfx90a");
+#endif
+#ifdef KOKKOS_ARCH_AMD_GFX940
+  check_architecture("gfx940");
+#endif
+#ifdef KOKKOS_ARCH_AMD_GFX942
+  check_architecture("gfx942");
+#endif
+#ifdef KOKKOS_ARCH_AMD_GFX1030
+  check_architecture("gfx1030");
+#endif
+#ifdef KOKKOS_ARCH_AMD_GFX1100
+  check_architecture("gfx1100");
+#endif
+#ifdef KOKKOS_ARCH_AMD_GFX1103
+  check_architecture("gfx1100");
+#endif
```

Comment:
How about an approach a la https://github.com/kokkos/kokkos/blob/daaee368fe4b8fbac079e0b398339b8573947b4c/core/src/impl/Kokkos_NvidiaGpuArchitectures.hpp#L17-L56

You define a `KOKKOS_IMPL_ARCH_AMD_GPU` macro as appropriate (integral value GCN arch or full arch name I haven't thought too much about it) which can be leveraged for other purposes as well.

Reply:
I would recommend you pick integral value in which case your sanity check comes down to
```C++
constexpr std::string_view required_arch =
    "gfx" KOKKOS_IMPL_TOSTRING(KOKKOS_IMPL_ARCH_AMD_GPU);
if (std::string_view detected_arch = Impl::HIPInternal::m_deviceProp.gcnArchName;
    detected_arch != required_arch) {
  // abort
}

Reply:
> You define a KOKKOS_IMPL_ARCH_AMD_GPU macro as appropriate 

I am not convinced about this. There is no equivalent to compute capabilities in HIP. You cannot check that you have at least a compute capability of 3.7 and then assume that some features are available. In HIP, there are macros for each individual feature. So I don't see how introducing this macro helps.

> I would recommend you pick integral value

That will not work for MI250 ( `gfx90a`)

Reply:
I think there is still value in having a macro that when defines denotes that we are targeting an AMD GPU.

Good point about `gfx90a`.

Reply:
We already have `KOKKOS_ARCH_AMD_GPU`

Reply:
Per our discussion at the office this morning let's define `KOKKOS_ARCH_AMD_GPU` as a GCN arch name literal string